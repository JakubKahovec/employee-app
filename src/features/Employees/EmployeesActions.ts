import { Employee } from "AppModel";
import { createAsyncAction } from "typesafe-actions";

export const saveEmployeeAsync = createAsyncAction(
  "employees/SAVE_EMPLOYEES_REQUEST",
  "employees/SAVE_EMPLOYEES_SUCCESS",
  "employees/SAVE_EMPLOYEES_FAILURE",
)<Employee, Employee, string>();

export const loadEmployeesAsync = createAsyncAction(
  "employees/LOAD_EMPLOYEES_REQUEST",
  "employees/LOAD_EMPLOYEES_SUCCESS",
  "employees/LOAD_EMPLOYEES_FAILURE",
)<undefined, Employee[], string>();
