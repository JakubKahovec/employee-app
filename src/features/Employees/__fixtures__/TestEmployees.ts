import { Employee } from "AppModel";

export const employee1: Employee = {
  jobTitle: "CEO",
  gender: "Male",
  tenure: "8",
  name: "John Green",
};

export const employee2: Employee = {
  jobTitle: "CTO",
  gender: "Female",
  tenure: "7",
  name: "Sheila Pink",
};

export const employee3: Employee = {
  jobTitle: "CFO",
  gender: "Female",
  tenure: "6",
  name: "Ema Shelton",
};

export const employee4: Employee = {
  jobTitle: "CFO",
  gender: "Male",
  tenure: "3",
  name: "Peter Holmes",
};
