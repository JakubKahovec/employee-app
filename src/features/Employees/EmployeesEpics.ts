import { Epic } from "redux-observable";
import { from, of } from "rxjs";
import { catchError, filter, map, mergeMap, switchMap } from "rxjs/operators";
import { isActionOf, RootAction, RootState, Services } from "typesafe-actions";

import { Employee } from "AppModel";
import { routerActions } from "connected-react-router";
import { routes } from "../../Routes";
import { loadEmployeesAsync, saveEmployeeAsync } from "./EmployeesActions";

export const saveEmployeeEpic: Epic<RootAction, RootAction, RootState, Services> = (action$, state$, { api }) =>
  action$.pipe(
    filter(isActionOf(saveEmployeeAsync.request)),
    mergeMap((action) =>
      from(api.employees.addEmployee(action.payload)).pipe(
        mergeMap((employee: Employee) => [saveEmployeeAsync.success(employee), routerActions.push(routes.dashboard)]),
        catchError((err: Error) => of(saveEmployeeAsync.failure(err.message))),
      ),
    ),
  );

export const loadEmployeesEpic: Epic<RootAction, RootAction, RootState, Services> = (action$, state$, { api }) =>
  action$.pipe(
    filter(isActionOf(loadEmployeesAsync.request)),
    switchMap(() =>
      from(api.employees.getEmployees()).pipe(
        map(loadEmployeesAsync.success),
        catchError((err: Error) => of(loadEmployeesAsync.failure(err.message))),
      ),
    ),
  );
