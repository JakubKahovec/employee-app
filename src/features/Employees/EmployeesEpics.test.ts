import { RootStateOrAny } from "react-redux";
import { ActionsObservable, StateObservable } from "redux-observable";
import { Subject } from "rxjs";
import { toArray } from "rxjs/operators";
import * as actions from "./EmployeesActions";
import { loadEmployeesEpic, saveEmployeeEpic } from "./EmployeesEpics";
import { employee1 } from "./__fixtures__/TestEmployees";

describe("employees epics", () => {
  it("should generate a success action when load employees api call is successful", async () => {
    const action$ = ActionsObservable.of(actions.loadEmployeesAsync.request());
    const state$ = new StateObservable(new Subject(), {} as RootStateOrAny);
    const epicResult$ = loadEmployeesEpic(action$, state$, {
      api: {
        employees: {
          getEmployees: jest.fn().mockResolvedValueOnce([employee1]),
          addEmployee: jest.fn().mockResolvedValue(employee1),
        },
      },
    });
    const resultValues = await epicResult$.pipe(toArray()).toPromise();
    expect(resultValues).toMatchSnapshot();
  });
  it("should generate a fail action when load employees api call fails", async () => {
    const action$ = ActionsObservable.of(actions.loadEmployeesAsync.request());
    const state$ = new StateObservable(new Subject(), {} as RootStateOrAny);
    const epicResult$ = loadEmployeesEpic(action$, state$, {
      api: {
        employees: {
          getEmployees: jest.fn().mockRejectedValue(new Error("Failed to call api")),
          addEmployee: jest.fn().mockResolvedValue(employee1),
        },
      },
    });
    const resultValues = await epicResult$.pipe(toArray()).toPromise();
    expect(resultValues).toMatchSnapshot();
  });
  it("should generate a success action and a redirect action when save employee api call is successful", async () => {
    const action$ = ActionsObservable.of(actions.saveEmployeeAsync.request(employee1));
    const state$ = new StateObservable(new Subject(), {} as RootStateOrAny);
    const epicResult$ = saveEmployeeEpic(action$, state$, {
      api: {
        employees: {
          getEmployees: jest.fn().mockResolvedValueOnce([employee1]),
          addEmployee: jest.fn().mockResolvedValue(employee1),
        },
      },
    });
    const resultValues = await epicResult$.pipe(toArray()).toPromise();
    expect(resultValues).toMatchSnapshot();
  });
  it("should generate a fail action when save employee api call fails", async () => {
    const action$ = ActionsObservable.of(actions.saveEmployeeAsync.request(employee1));
    const state$ = new StateObservable(new Subject(), {} as RootStateOrAny);
    const epicResult$ = saveEmployeeEpic(action$, state$, {
      api: {
        employees: {
          getEmployees: jest.fn().mockResolvedValueOnce([employee1]),
          addEmployee: jest.fn().mockRejectedValue(new Error("Failed to call api")),
        },
      },
    });
    const resultValues = await epicResult$.pipe(toArray()).toPromise();
    expect(resultValues).toMatchSnapshot();
  });
});
