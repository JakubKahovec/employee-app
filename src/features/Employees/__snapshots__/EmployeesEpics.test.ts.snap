// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`employees epics should generate a fail action when load employees api call fails 1`] = `
Array [
  Object {
    "payload": "Failed to call api",
    "type": "employees/LOAD_EMPLOYEES_FAILURE",
  },
]
`;

exports[`employees epics should generate a fail action when save employee api call fails 1`] = `
Array [
  Object {
    "payload": "Failed to call api",
    "type": "employees/SAVE_EMPLOYEES_FAILURE",
  },
]
`;

exports[`employees epics should generate a success action and a redirect action when save employee api call is successful 1`] = `
Array [
  Object {
    "payload": Object {
      "gender": "Male",
      "jobTitle": "CEO",
      "name": "John Green",
      "tenure": "8",
    },
    "type": "employees/SAVE_EMPLOYEES_SUCCESS",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/dashboard",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;

exports[`employees epics should generate a success action when load employees api call is successful 1`] = `
Array [
  Object {
    "meta": 0,
    "payload": Array [
      Object {
        "gender": "Male",
        "jobTitle": "CEO",
        "name": "John Green",
        "tenure": "8",
      },
    ],
    "type": "employees/LOAD_EMPLOYEES_SUCCESS",
  },
]
`;
